### PR TITLE
python: add record size limit

### PR DESCRIPTION
**Pull request by james-rms**
### Changelog
- Added: the python MCAP library reader classes gained a new parameter, `record_size_limit`, defaulting to 4GiB. This allows callers to limit the size of records that their application will support. When a record is encountered with a greater length, the reader will raise an `InvalidRecordLength` exception. This limit can be removed by setting it to `None`. This helps applications avoid the issue where a corrupt MCAP can cause a `MemoryError`.

### Docs

See generated python docs in the mcap.dev preview.
### Description

<table><tr><th>Before</th><th>After</th></tr><tr><td>

</td><td>

</td></tr></table>

Fixes: #1220

### CODE REVIEW

File: python/mcap/mcap/exceptions.py [hunk 1]
```diff
@@ -13,3 +18,10 @@ class DecoderNotFoundError(McapError):
 
 class EndOfFile(McapError):
     pass
+
+
+class InvalidRecordLength(McapError):
+    def __init__(self, opcode: Opcode, length: int, limit: int):
+        super().__init__(
+            f"{opcode.name} record has invalid length {length}, limit is set to {limit}",
+        )
```
Comment (indirectlylit) [inline python/mcap/mcap/exceptions.py:27]: Is the record length actually invalid, or just too big?

```suggestion
class RecordLengthLimitExceeded(McapError):
    def __init__(self, opcode: Opcode, length: int, limit: int):
        super().__init__(
            f"{opcode.name} record has length {length}, limit is set to {limit}",
        )
```

File: python/mcap/mcap/reader.py [hunk 2]
```diff
@@ -240,19 +240,26 @@ class SeekingReader(McapReader):
     :param decoder_factories: An iterable of :py:class:`~mcap.decoder.DecoderFactory`
         instances which can provide decoding functionality to
         :py:meth:`~mcap.reader.McapReader.iter_decoded_messages`.
+    :param record_size_limit: An upper bound to the size of MCAP records that this reader will
+        attempt to load, defaulting to 4 GiB. If this reader encounters a record with a greater
```
Comment (indirectlylit) [inline python/mcap/mcap/reader.py:244]: if this is an integer number of bytes, maybe specify that detail?

**Review by indirectlylit — COMMENTED** [review]
Seems like a good sanity check.

I tried testing with the example in linear, and was expecting to see the new exception but instead saw `ValueError: 51 is not a valid Opcode`.

Before:

```
python test.py
Traceback (most recent call last):
  File "test.py", line 3, in <module>
    for msg in read_ros2_messages("minimal-mcap-to-trigger-memory-error.mcap"):
  File "/Users/d/Projects/mcap/python/mcap-ros2-support/mcap_ros2/reader.py", line 74, in read_ros2_messages
    for schema, channel, message, ros2_msg in reader.iter_decoded_messages(
  File "/Users/d/Projects/mcap/python/mcap/mcap/reader.py", line 191, in iter_decoded_messages
    for schema, channel, message in message_iterator:
  File "/Users/d/Projects/mcap/python/mcap/mcap/reader.py", line 278, in iter_messages
    summary = self.get_summary()
  File "/Users/d/Projects/mcap/python/mcap/mcap/reader.py", line 338, in get_summary
    footer = next(StreamReader(self._stream, skip_magic=True).records)
  File "/Users/d/Projects/mcap/python/mcap/mcap/stream_reader.py", line 120, in records
    record = self._read_record(opcode, length)
  File "/Users/d/Projects/mcap/python/mcap/mcap/stream_reader.py", line 205, in _read_record
    self._stream.read(length - 9)
  File "/Users/d/Projects/mcap/python/mcap/mcap/data_stream.py", line 26, in read
    data = self._stream.read(length)
MemoryError
```

after:

```
python test.py                                                       
Traceback (most recent call last):
  File "test.py", line 3, in <module>
    for msg in read_ros2_messages("minimal-mcap-to-trigger-memory-error.mcap"):
  File "/Users/d/Projects/mcap/python/mcap-ros2-support/mcap_ros2/reader.py", line 74, in read_ros2_messages
    for schema, channel, message, ros2_msg in reader.iter_decoded_messages(
  File "/Users/d/Projects/mcap/python/mcap/mcap/reader.py", line 191, in iter_decoded_messages
    for schema, channel, message in message_iterator:
  File "/Users/d/Projects/mcap/python/mcap/mcap/reader.py", line 285, in iter_messages
    summary = self.get_summary()
  File "/Users/d/Projects/mcap/python/mcap/mcap/reader.py", line 351, in get_summary
    footer = next(
  File "/Users/d/Projects/mcap/python/mcap/mcap/stream_reader.py", line 160, in records
    Opcode(opcode), length, self._record_size_limit
  File "/Users/d/.pyenv/versions/3.8.18/lib/python3.8/enum.py", line 339, in __call__
    return cls.__new__(cls, value)
  File "/Users/d/.pyenv/versions/3.8.18/lib/python3.8/enum.py", line 663, in __new__
    raise ve_exc
ValueError: 51 is not a valid Opcode
```

is that expected?

**Comment by james-rms** [comment]
@indirectlylit nope updated. Thanks for the review!

File: python/mcap/mcap/exceptions.py [hunk 3]
```diff
@@ -13,3 +18,16 @@ class DecoderNotFoundError(McapError):
...
+    def __init__(self, opcode: int, length: int, limit: int):
+        opcode_name = f"unknown (opcode {opcode})"
+        try:
+            opcode_name = Opcode(opcode).name
+        except ValueError:
+            # unknown opcode will trigger a ValueError
+            pass
+        super().__init__(
+            f"{opcode_name} record has length {length} that exceeds limit {limit}",
+        )
```
Comment (indirectlylit) [inline python/mcap/mcap/exceptions.py:33]: Nit: could possibly tighten up the language a bit. E.g. from

`unknown (opcode 51) record has length 4049071657446291766 that exceeds limit 4294967296`

to 

`Unknown record (opcode 51) has length 4049071657446291766, which exceeds limit 4294967296`


```suggestion
        record_name = None
        try:
            record_name = f"{Opcode(opcode).name} record"
        except ValueError:
            # unknown opcode will trigger a ValueError
            record_name = f"Unknown record (opcode {opcode})"
        super().__init__(
            f"{record_name} has length {length} that exceeds limit {limit}",
        )
```

**Review by indirectlylit — APPROVED** [review]
looks good! new output:

```
python test.py
Traceback (most recent call last):
  File "test.py", line 3, in <module>
    for msg in read_ros2_messages("minimal-mcap-to-trigger-memory-error.mcap"):
  File "/Users/d/Projects/mcap/python/mcap-ros2-support/mcap_ros2/reader.py", line 74, in read_ros2_messages
    for schema, channel, message, ros2_msg in reader.iter_decoded_messages(
  File "/Users/d/Projects/mcap/python/mcap/mcap/reader.py", line 191, in iter_decoded_messages
    for schema, channel, message in message_iterator:
  File "/Users/d/Projects/mcap/python/mcap/mcap/reader.py", line 285, in iter_messages
    summary = self.get_summary()
  File "/Users/d/Projects/mcap/python/mcap/mcap/reader.py", line 351, in get_summary
    footer = next(
  File "/Users/d/Projects/mcap/python/mcap/mcap/stream_reader.py", line 159, in records
    raise RecordLengthLimitExceeded(opcode, length, self._record_size_limit)
mcap.exceptions.RecordLengthLimitExceeded: unknown (opcode 51) record has length 4049071657446291766 that exceeds limit 4294967296
```

**Review by achim-k — APPROVED** [review]

**Pull request merged**